pkg/frontend/querymiddleware/cardinality.go

```diff
@@ -0,0 +1,155 @@
+		return &cardinalityEstimation{
+			cache:  cache,
+			next:   next,
+			logger: logger,
+
+			cardinalityEstimationMetrics: metrics,
+		}
+	})
+}
+
+// Do injects a cardinality estimate into the query hints (if available) and
+// caches the actual cardinality observed for this query.
+func (c *cardinalityEstimation) Do(ctx context.Context, request Request) (Response, error) {
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, c.logger, "cardinalityEstimation.Do")
+	defer spanLog.Finish()
+
+	tenants, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return c.next.Do(ctx, request)
+	}
+
+	k := generateCardinalityEstimationCacheKey(tenant.JoinTenantIDs(tenants), request, cardinalityEstimateBucketSize)
+
+	var estimatedCardinality uint64
+	if estimate, ok := c.lookupCardinalityForKey(ctx, k); ok {
+		estimatedCardinality = estimate
+		request = request.WithEstimatedCardinalityHint(estimate)
+	}
+
+	res, err := c.next.Do(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	statistics := stats.FromContext(ctx)
+	actualCardinality := statistics.GetFetchedSeriesCount()
+
+	if !isCardinalitySimilar(actualCardinality, estimatedCardinality) {
+		c.storeCardinalityForKey(ctx, k, actualCardinality)
+	}
+
+	if estimatedCardinality > 0 {
+		c.estimationError.Observe(math.Abs(float64(actualCardinality) - float64(estimatedCardinality)))
+		statistics.AddEstimatedSeriesCount(estimatedCardinality)
+		spanLog.LogKV("estimated cardinality", estimatedCardinality, "actual cardinality", actualCardinality)
+	}
+
+	return res, nil
+}
+
+// lookupCardinalityForKey fetches a cardinality estimate for the given key from
+// the results cache.
+func (c *cardinalityEstimation) lookupCardinalityForKey(ctx context.Context, key string) (uint64, bool) {
+	if c.cache == nil {
+		return 0, false
+	}
+	res := c.cache.Fetch(ctx, []string{key})
+	if val, ok := res[key]; ok {
+		qs := &QueryStatistics{}
+		err := proto.Unmarshal(val, qs)
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "failed to unmarshal cardinality estimate")
+			return 0, false
+		}
+		return qs.EstimatedSeriesCount, true
+	}
+	return 0, false
+}
+
+// storeCardinalityForKey stores a cardinality estimate for the given key in the
+// results cache.
+func (c *cardinalityEstimation) storeCardinalityForKey(ctx context.Context, key string, count uint64) {
+	if c.cache == nil {
+		return
+	}
+	m := &QueryStatistics{EstimatedSeriesCount: count}
+	marshaled, err := proto.Marshal(m)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to marshal cardinality estimate")
+	}
+	// The store is executed asynchronously, potential errors are logged and not
+	// propagated back up the stack.
+	c.cache.Store(ctx, map[string][]byte{key: marshaled}, cardinalityEstimateTTL)
+}
+
+func isCardinalitySimilar(actualCardinality, estimatedCardinality uint64) bool {
+	estimate := float64(estimatedCardinality)
+	actual := float64(actualCardinality)
+	res := math.Abs(estimate/actual-1) < cacheErrorToleranceFraction
```

Comment:
Is the `actual-1` to protect from a division by zero? You still have a division by 0 when actual is 1.

Reply:
No, it's just how the formula works out. Good catch on this, I hadn't considered that it's legit to have an actual count of 0.